## Problem : Perform complete logistic regression analysis of Bankruptcy Data

# Step 1 : Data Preparation

In [1]:
import pandas as pd
df=pd.read_csv('bankruptcy.csv')
X=df.iloc[:,:4]
y=df.iloc[:,4]

# Step 2: Fitting the Logistic Regression Model

In [2]:
import statsmodels.api as sm
x=sm.add_constant(X)
model=sm.Logit(y,x).fit()
model.summary()

Optimization terminated successfully.
         Current function value: 0.298296
         Iterations 8


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                     cl   No. Observations:                   46
Model:                          Logit   Df Residuals:                       41
Method:                           MLE   Df Model:                            4
Date:                Sun, 22 Oct 2023   Pseudo R-squ.:                  0.5673
Time:                        20:33:34   Log-Likelihood:                -13.722
converged:                       True   LL-Null:                       -31.711
Covariance Type:            nonrobust   LLR p-value:                 2.924e-07
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -5.3195      2.366     -2.248      0.025      -9.958      -0.681
X1             7.1378      6.002      1.189      0.234      -4.625      18.901
X2            -3.7033     13.670     -0.271      0.786     -30.497      23.090
X3             3.4148      1.204      2.837      0.005       1.056       5.774
X4            -2.9684      3.065     -0.968      0.333      -8.976       3.040
==============================================================================
"""

# Results: <br>
## The fitted model is $y=\frac{1}{1+e^{5.3195-7.1378x_1+3.1378x_2-3.4148x_3+2.9684x_4}}$<br>
## The estimated values are $\hat\beta_0=-5.3195,\hat\beta_1=,\hat\beta_2=,\hat\beta_3=,\hat\beta_4= $
## The most significant regressor is   $x_3$
## The most insignificant regressor is $x_2$

# Odds Ratios

In [3]:
print('Estimated values are', model.params)

Estimated values are const   -5.319513
X1       7.137804
X2      -3.703330
X3       3.414834
X4      -2.968390
dtype: float64


In [4]:
import numpy as np
print('Odds Ratios are ', np.exp(model.params))

Odds Ratios are  const       0.004895
X1       1258.661658
X2          0.024641
X3         30.411911
X4          0.051386
dtype: float64


# Odd Ratio of the regressor $x_3$ is unusually large 

# Inline prediction

In [5]:
y_pr=np.round(model.predict(x))


## From the Pseudo R-Square we conclude that the model is a good model